feat: add a utility function that determines whether a part has slices

linostar opened the pull request:
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
A utility function is needed in rockcraft to determine whether parts has slices in them. 
Instead of implementing such a function in rockcraft, it is better to implement it in craft-parts instead.

linostar commented:
@tigarmo Ready for review.

linostar commented:
Tagging @cmatsuoka @lengau @mr-cal for review.

Failing integreation test is due to network connection issue so it needs a retrigger.

cmatsuoka commented on craft_parts/parts.py:
```diff
@@ -642,6 +639,23 @@ def part_has_overlay(data: Dict[str, Any]) -> bool:
     return spec.has_overlay
 
 
+def part_has_slices(data: Dict[str, Any]) -> bool:
+    """Whether the part described by ``data`` has slices in its stage-packages.
```
How and when is this going to be accessed? It feels that it should be a part property instead of a call receiving a list of part names.

linostar replied:
This is needed for Rockcraft's [PR 464](https://github.com/canonical/rockcraft/pull/464).
It will work as a part property as well, and I don't have a particular preference for either. However, there is already a similar utility function called `part_has_overlay`, so I put this one in the same scope.

tigarmo commented on craft_parts/parts.py:
```diff
@@ -642,6 +639,23 @@ def part_has_overlay(data: Dict[str, Any]) -> bool:
     return spec.has_overlay
 
 
+def part_has_slices(data: Dict[str, Any]) -> bool:
```
I think it's important to add a basic test for `part_has_slices()` and `is_slice()`. They are currently covered by the existing validator test but if refactor the validator the functions might become uncovered

linostar commented:
I have changed the utility function to a property, and added unit tests for it. This is to accomodate the function to how similar functions are unit tested.
Note the failing unit-tests checks are not related to this PR.

lengau commented on craft_parts/parts.py:
```diff
@@ -169,6 +166,15 @@ def has_overlay(self) -> bool:
+        :param data: The part data to query.
+        """
+        stage_packages = self.stage_packages or []
+        return any(name for name in stage_packages if is_slice(name))
```
```suggestion
        return any(is_slice(name) for name in stage_packages)
```

lengau reviewed (approved):
Code looks good, thanks :-)

lengau commented:
@linostar could you rebase this on main and resolve the conflicts here?

cmatsuoka commented on craft_parts/parts.py:
```diff
@@ -175,6 +172,15 @@ def has_overlay(self) -> bool:
+
+        :param data: The part data to query.
+        """
+        stage_packages = self.stage_packages or []
```
I would test for `self.stage_packages` and short-circuit the return value here.

cmatsuoka reviewed (commented):
LGTM, maybe we could make it slightly more performant with a short-circuit return in `has_slice`.

linostar commented:
@cmatsuoka @tigarmo Can we merge this PR if there are no further comments, so that we can move forward with https://github.com/canonical/rockcraft/pull/464?

linostar commented:
This is closed in favor of #904.
